Compute `target_feature` from LLVM

ranma42 opened the pull request:
This is a work-in-progress fix for #31662.

The logic that computes the target features from the command line has been replaced with queries to the `TargetMachine`.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

ranma42 commented:
I tried this and it has the desired result for the example of #31662, but I think it might need more work.
In particular, I am not confident about several issues:
 - is this an acceptable approach?
 - is a string the desired output type?
 - is it ok to work on a `TargetMachine` instance?
 - would it be ok to initialise LLVM earlier than in trans? (would it be ok to do it from `build_configuration`?)

I do not think this is ready for merge, but I would love feedback and directions.

alexcrichton commented on src/librustc_driver/target_features.rs:
```diff
@@ -24,64 +24,14 @@ use syntax::parse::token::intern_and_get_ident as intern;
-        has_avx, "avx";
-        has_avx2, "avx2";
-        has_neon, "neon";
-        has_vfp, "vfp";
```
One of the purposes of a macro like this was to ensure that we had control over the names exported. We don't want to blanket export all features that LLVM itself exposes, and we also don't want to tie ourselves to exactly the names LLVM has. In other words this is here to ensure we can provide a stable interface over the possibly unstable interface of LLVM.

ranma42 replied:
Filtering and renaming these features to a "rust" convention instead of the LLVM one can be done very easily with a map. The filtering makes sense to me, as rust might want to expose only a subset of the LLVM features through `cfg!`.
Renaming these features is just as easy, but I would be surprised if it happened only here and not in every place where a feature is named (for example when they are passed through the `-C target-feature` flag). That would result in an inconsistent naming, as the same feature would have a different name in the command line flags and in the code.
If you confirm that renaming is desired here and not elsewhere, I will implement it, otherwise I believe that it might belong to another PR (under the assumption that if LLVM and rust feature naming was to diverge in future, we would decide how to take care of that and possibly perform a consistent renaming everywhere).

alexcrichton replied:
The purpose of this macro was that for the function/name pair the cfg directive name would be defined if the function returned true. This is the only place that does that.

Basically, we need to filter what's coming out of LLVM with an explicit whitelist of what cfg directives are defined. We don't need to map names yet (as we just use LLVM names), but we may support that one day.

ranma42 replied:
I added a whitelist of known LLVM features which is used to filter which ones are exposed through `cfg!`.

alexcrichton commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -97,6 +97,55 @@ LLVMRustAddPass(LLVMPassManagerRef PM, Pass *pass) {
+
+#define SUBTARGET(x) namespace llvm {                \
+    extern const SubtargetFeatureKV x##FeatureKV[];  \
+    extern const SubtargetFeatureKV x##SubTypeKV[];  \
```
Is this just reaching into the internals of LLVM without including header files? That seems... brittle?

ranma42 replied:
These symbols are public, but unfortunately they are not exposed through any of the public headers. I will try to ask in the LLVM mailing list about a better way to access the feature names, but this was the best option I could find among the currently available alternatives.

alexcrichton replied:
I'd be fine doing this for already-released LLVM versions, but I'd much prefer to get a patch upstream in LLVM exposing these symbols in header files so the compiler can check the type definitions.

alexcrichton commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -97,6 +97,55 @@ LLVMRustAddPass(LLVMPassManagerRef PM, Pass *pass) {
+    SUBTARGET(ARM) \
+    SUBTARGET(AArch64) \
+    SUBTARGET(Mips) \
+    SUBTARGET(PPC)
```
Support for each subtarget is conditional most of the time because not all LLVMs are built with all these targets enabled, so this'll need to take that into account somehow

ranma42 replied:
The LLVM components detected at build time are available through `cfg!(llvm_component)` in Rust, but I could not find anything similar for the C++ libraries. Is it ok to duplicate the detection code from `mk/llvm.mk`?

alexcrichton replied:
Let's avoid having *too* much duplication, but otherwise yes I suspect something similar will have to happen here

alexcrichton commented on src/librustc_driver/target_features.rs:
```diff
@@ -24,64 +24,14 @@ use syntax::parse::token::intern_and_get_ident as intern;
-    }
-}
-
+    unsafe { ::rustc_trans::back::write::configure_llvm(sess); }
```
I think there's a `Once` guarding this somewhere else, and this probably wants to go that same route rather than running this unconditionally (which can be racy)

ranma42 replied:
What would be the best time to perform the LLVM initialisation? Would it be ok to move it to `build_configuration`?
Assuming that `build_configuration` exposes the target features and that they are computed by LLVM, the initialization cannot happen later than that.

alexcrichton replied:
I think it's fine to basically initialize LLVM whenever, but it needs to happen only once. This just needs to share the same `Once` that trans does (perhaps by moving it into literally this method being called or something like that)

nikomatsakis commented:
r? @alexcrichton

ranma42 commented:
I moved the initialisation of LLVM to librustc. In order to avoid a circular dependency between crates, I had to move `configure_llvm`. This might not be correct, as it was a public function. Currently I made it private, but if it is meant to be part of the public interface of `librustc_trans` it might need to be reexported or moved to a crate which can be used/exported by `librustc` and `librustc_trans`.

alexcrichton commented on src/librustc/session/config.rs:
```diff
@@ -720,6 +750,27 @@ pub fn build_configuration(sess: &Session) -> ast::CrateConfig {
     v.extend_from_slice(&default_cfg[..]);
+
+    // Before we touch LLVM, make sure that multithreading is enabled.
+    unsafe {
```
This should be extracted into its own function instead of being copy/pasted into an otherwise irrelevant function. This should also probably be called from a more principled location such as the driver itself

ranma42 replied:
You mean in `librustc_driver`? AFAICT it is not possible to invoke functions in `librustc_driver` from `librustc` because the crates would have a circular dependency. Should I also try to move `build_configuration` to another crate? (is it possible/does it make sense?)

alexcrichton replied:
Yeah, it'd just be an implicit contract that the driver must initialize LLVM before calling this function.

alexcrichton commented on src/librustc/session/config.rs:
```diff
@@ -709,6 +709,36 @@ pub fn append_configuration(cfg: &mut ast::CrateConfig,
     }
 }
 
+unsafe fn configure_llvm(sess: &Session) {
```
Let's keep this where it lived initially and perhaps just call it from the driver instead

ranma42 replied:
It would be extremely convenient if it was possible to call it from some kind of entry point in `librustc_driver` (instead of `build_configuration`) but I was unable to identify such function.

alexcrichton replied:
If one doesn't exist, can one be made? I'd be fine just documenting that things won't work unless some initialization function is called first.

alexcrichton commented on src/librustc_driver/target_features.rs:
```diff
@@ -24,64 +24,20 @@ use syntax::parse::token::intern_and_get_ident as intern;
-        has_vfp, "vfp";
-    }
-}
+    let target_machine = ::rustc_trans::back::write::create_target_machine(sess); 
```
Let's use imports here instead of hacking around with `::` paths (for this and many of the lines below).

Also, can you be sure to wrap lines at 80 chars? (the surrounding style)

ranma42 commented:
I tried to implement all of the suggested changes. It might be possible that PNaCl needs special care if it has features we want to expose, because it does not look like a normal LLVM target. So far this should not be an issue, as all of the features that are whitelisted are from other targets.

I added the `init_llvm` function in `librustc_trans::back::write` and it should be called before any use of LLVM. Is `src/librustc/README.md` the most appropriate place to warn people that `init_llvm` should be called?

alexcrichton commented:
Unfortunately I think the makefiles will also need to change to account for the change here, rustbuild is only an alternate build system currently. It's ok to not worry much about pnacl for now though.

And yeah, `init_llvm` looks good to me. It's probably fine to leave it sprinkled as-is, it's not clear to me anyway about who's supposed to call it anyway.

ranma42 commented:
I added the changes to Makefiles (based on the existing build templates from mk/main.mk and  mk/llvm.mk) and fixed the tests that were using LLVM to invoke `init_llvm` as part of the session construction.

alexcrichton commented on mk/rustllvm.mk:
```diff
@@ -43,6 +43,10 @@ $$(RT_OUTPUT_DIR_$(1))/$$(call CFG_STATIC_LIB_NAME_$(1),rustllvm): \
 
+EXTRA_RUSTLLVM_CXXFLAGS_$(1) = $$(shell echo $$(filter $$(LLVM_ALL_COMPONENTS_$(1)),\
+	$$(LLVM_SUBTARGET_COMPONENTS)) | tr 'a-z' 'A-Z' |\
+	sed -e 's/^ //;s/\([^ ]*\)/\-DLLVM_SUBTARGET_\1/g')
```
Perhaps this could just be:

```make
EXTRA_RUSTLLVM_CXXFLAGS_$(1) = $$(LLVM_ALL_COMPONENTS_$(1):%=-DLLVM_COMPONENT_%)
```

That way we could just ferry all components down to the C++ code?

ranma42 replied:
Oops, you just made me realise that I should not be overwriting `EXTRA_RUSTLLVM_CXXFLAGS_$(1)`, I meant to use the `RUSTLLVM_TARGETS_$(1)` variable!
Anyway, yes, that would work and looks much simpler, but it will result in mixed-case defines. Is it ok?

ranma42 replied:
Hmm... it does not work, because the `-` character is not allowed in a define, but it is used in several components. Something like this seems to work:
```
RUSTLLVM_COMPONENTS_$(1) = $$(shell echo $$(LLVM_ALL_COMPONENTS_$(1)) |\
        tr 'a-z-' 'A-Z_'| sed -e 's/^ //;s/\([^ ]*\)/\-DLLVM_COMPONENT_\1/g')
```
and still avoids the distinction between target and non-target components.

alexcrichton commented on src/librustdoc/core.rs:
```diff
@@ -129,6 +129,7 @@ pub fn run_core(search_paths: SearchPaths, cfgs: Vec<String>, externs: Externs,
     let sess = session::build_session_(sessopts, cpath, diagnostic_handler,
                                        codemap, cstore.clone());
     rustc_lint::register_builtins(&mut sess.lint_store.borrow_mut(), Some(&sess));
+    write::init_llvm(&sess);
```
Hm ok this may be a *little* too much sprinkling around. Perhaps this could just be added to the constructor of a `Session`?

ranma42 replied:
I would prefer it a lot (if you remember, it was my original approach), but AFAICT it requires moving the `init_llvm` code in order to avoid cyclic dependencies between `librustc` (where `Session` is constructed) and `librustc_trans`. Is that acceptable? Otherwise, is there a clever way to workaround the cyclic dependency without moving the code? (actually I would like it if it was not too clever... I would hate it to be unintuitive)

ranma42 commented:
@alexcrichton I tried to implement both your suggestions and the overall change now looks cleaner :)
I kept the new commits separate, but it might be a good idea to squash them before merging so that the changes that have been reverted (those to the tests, for example) are cleaned up.

alexcrichton commented on src/librustc_llvm/build.rs:
```diff
@@ -94,6 +94,13 @@ fn main() {
+    for component in components {
+        let mut flag = String::from("-DLLVM_COMPONENT_");
+        flag.push_str(&component.to_uppercase());
+        cfg.flag(&flag);
```
zomg this looks so much nicer than make

alexcrichton commented:
Awesome, looking great to me, thanks @ranma42!

Could you add some tests for this as well? I think the case in #31662 may be a good one to add?

alexcrichton commented:
Also yeah I'd be fine squashing commits before landing as well.

alexcrichton commented:
Also, out of curiosity, did you get a response from asking LLVM [about these constants](https://github.com/rust-lang/rust/pull/31709#discussion_r53097030)?

ranma42 commented:
I tried to adapt the program from #31662, but it is not trivial to make a test which is correct for all possible targets.  AFAICT the idea is that the test should check whether the `sse` feature is available on `i686-unknown-linux-gnu`. How can I write a test that is only run on a specific triple? I would like to avoid having the test broken on `i586-unknown-linux-gnu`.

Would a `print-cfg` test like the following be sufficient?
```
$(RUSTC) --target i686-unknown-linux-gnu --print cfg | grep sse2
```
This test would not pass on stable/beta release channels until `target_feature` is stabilised. Is this a problem?

This morning I sent a message to the llvm-dev mailing list pointing out this approach for computing the features and asking for guidance. The relevant thread starts [here](http://lists.llvm.org/pipermail/llvm-dev/2016-March/096480.html).

alexcrichton commented:
Ah yeah a test like that should be fine. We've export the `RUST_BOOTSTRAP_KEY` during the CI builds of the stable/beta channels so I *think* that will pass even for those channels?

alexcrichton commented:
Also thanks for starting that thread! It's probably worth waiting a little bit to see if anyone responds.

bors commented:
:umbrella: The latest upstream changes (presumably #32432) made this pull request unmergeable. Please resolve the merge conflicts.

ranma42 commented:
rebased

alexcrichton commented:
@bors: r+ 88e26f3302802d93b60526fb4e139edef0abcb61

Thanks @ranma42!

bors commented:
:hourglass: Testing commit 88e26f3 with merge f679e0e...

bors commented:
:snowman: The build was interrupted to prioritize another pull request.

bors commented:
:hourglass: Testing commit 88e26f3 with merge 41208fb...

bors commented:
:snowman: The build was interrupted to prioritize another pull request.

bors commented:
:hourglass: Testing commit 88e26f3 with merge 86beac0...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3619)

alexcrichton commented:
@bors: retry

On Wed, Mar 30, 2016 at 11:28 AM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-gnu-32-nopt-t
> <http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3619>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31709#issuecomment-203568519>
>

bors commented:
:broken_heart: Test failed - [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1764)

bors commented:
:umbrella: The latest upstream changes (presumably #32628) made this pull request unmergeable. Please resolve the merge conflicts.

ranma42 commented:
rebased

alexcrichton commented:
@bors: r+ 325ea9529dc4ce9c8c64a34b44af6d6f8802ab52

bors commented:
:hourglass: Testing commit 325ea95 with merge 7017863...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3611)

alexcrichton commented:
@bors: retry

On Thu, Mar 31, 2016 at 2:08 PM, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-gnu-64-opt
> <http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3611>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31709#issuecomment-204130495>
>

bors commented:
:hourglass: Testing commit 325ea95 with merge 489e9e5...

bors commented:
:broken_heart: Test failed - [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/1771)

bors commented:
:umbrella: The latest upstream changes (presumably #32562) made this pull request unmergeable. Please resolve the merge conflicts.

ranma42 commented:
Rebased. Is the failure just noise or an actual one? The log does not provide much information about what has gone wrong :(

alexcrichton commented:
I don't think that this will get fixed from a rebase, the multiple segfaults both seem legitimate.

Both happen when compiling the PowerPC libcore which seems to indicate that the `extern` definitions here may not be correct for PowerPC?

bors commented:
:umbrella: The latest upstream changes (presumably #32767) made this pull request unmergeable. Please resolve the merge conflicts.

ranma42 commented:
The failure was caused by some of the elements of the `featureBits` not being listed as actual features. I changed the approach so that only the whitelisted features are queried.

alexcrichton commented on src/librustc_driver/target_features.rs:
```diff
@@ -9,79 +9,47 @@
+    let tf = InternedString::new("target_feature");
+    for feat in whitelist {
+        if unsafe { LLVMRustHasFeature(target_machine, feat.as_ptr() as *const c_char) } {
+            cfg.push(attr::mk_name_value_item_str(tf.clone(), intern(feat)))
```
This interns the option with a trailing nul, right? I think that may prevent matching in Rust code?

ranma42 replied:
Right, I should be interning the string with nun sliced away.

alexcrichton commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -97,6 +97,75 @@ LLVMRustAddPass(LLVMPassManagerRef PM, Pass *pass) {
+#undef SUBTARGET
+
+    while (strcmp(feature, FeatureEntry->Key) != 0)
+        FeatureEntry++;
```
This seems kind dangerous like it's easy to walk of the end of an array, does NULL or something like that indicate "we've reached the end"?

ranma42 replied:
Unfortunately no such terminator is available :(
The previous version failed on PPC exactly for this reason: some of the features bits used on the PPC CPUs are not listed in the corresponding feature array, hence the linear search went past the end of the array without finding a match.
Under the assumption that the whitelist only contains items that are in the feature array, this loop should never walk past the end.

alexcrichton replied:
Right! Man that's crazy...

huonw replied:
Seems like we could count/compare the number of things we've looked at with the size of the bitset (do we know that)? I.e. to protect against that danger.

alexcrichton commented:
Looks good to me! Can a test also be added that exercises `#[cfg(target_feature)]` itself?

ranma42 commented:
As mentioned [here](https://github.com/rust-lang/rust/pull/31709#issuecomment-193717807) it would be hard to make such test work on different targets.
Do you have any suggestion on how that could be best accomplished?

tamird commented on src/librustc/session/mod.rs:
```diff
@@ -491,9 +494,63 @@ pub fn build_session_(sopts: config::Options,
+        if sess.time_llvm_passes() { add("-time-passes"); }
+        if sess.print_llvm_passes() { add("-debug-pass=Structure"); }
+
+        // FIXME #21627 disable faulty FastISel on AArch64 (even for -O0)
```
looks like this is safe to remove now; https://github.com/rust-lang/rust/issues/21627 references an upstream fix which went into llvm 18 months ago https://github.com/llvm-mirror/llvm/commit/ca07e256f62f772d14c42f41af46f2aeacc54983

alexcrichton replied:
Awesome! Would you be interested in prep'ing a separate PR for that?

tamird replied:
sure. I'll open a new PR if that removal doesn't happen here.

alexcrichton replied:
Yeah it's probably good to keep these separate, so let's split that change out of this PR.

alexcrichton commented:
Oh right, yeah, sorry I forgot about that! Let's actually go ahead and check in an x86-specific test and if we ever get the test suite running on i586 we can just add an `// ignore-i586` directive perhaps?

ranma42 commented on src/librustc_driver/target_features.rs:
```diff
@@ -9,79 +9,47 @@
 
+    let arm_whitelist = [
+        "neon\0",
+        "vfp\0",
```
Which LLVM feature corresponds to the `vfp` one listed here?
LLVM distinguishes between `vfp2`, `vfp3` and `vfp4`. Each one implies the lower versions and `neon` implies `vfp3`.
For some additional information, see https://wiki.debian.org/ArmHardFloatPort/VfpComparison

alexcrichton replied:
Looks like gcc has options like vfp, vfpv3, vfpv3d16, and vfpv4. Also looks like debian says vfpv1 is deprecated, so I guess vfpv2?

ranma42 replied:
Right, given the gcc precedent, that seems likely.
So far this is the only instance of an LLVM feature which is exposed under a different name in Rust. This would mean that the feature can be enabled with `-C target-feature=+vfp[234]`, but that it exposed as `cfg!(target_feature="vfp")`. Is it intentional? In this case, I will extend the whitelist logic to also perform renaming. Otherwise it would be possible to expose it (and/or the other `vfp` versions) as separate features.

alexcrichton replied:
Whether anything here is intentional I'm not sure. I believe @huonw originally wrote this in, but he's out for this week and may take a moment to respond. I suspect that this was just filling in features on ARM, however.

To me it makes sense that enabling vfp4, however, would set the vfp `target_feature` because, for example, enabling `avx` enables a bunch of other features (like `sse2`).

ranma42 replied:
I will check that everything else is ready and wait for @huonw.

Since LLVM tracks feature dependencies, enabling `vfp4` would enable `vfp3` (which in turn would also enable `vfp2`), so `-C target_feature=+vfp4` would enable all of them on the LLVM side. My question is whether `+vfp4` on the LLVM side should enable the `vfp` feature in Rust or if it should rather enable the `vfp4` feature (as well as others that are implied).

alexcrichton replied:
FWIW this is unstable, and @huonw's [simd repo](https://github.com/huonw/simd) doesn't mention vfp, so it's probably fine to change. In light of this information it seems most reasonable to follow what LLVM is doing with vfp{2,3,4}

ranma42 commented:
I fixed the NUL string terminator issue, added a comment about the risk of going past the array bounds in LLVM and exposed each of the `vfp?` features independently. I also added a test which checks that sse2 is available on x86 archs.

I did not include the cleanup mentioned by @tamird .

alexcrichton commented:
@bors: r+ 1ad85610e4eae8b7281c85186dff8edb23b5d21c

Thanks @ranma42!

bors commented:
:hourglass: Testing commit 1ad8561 with merge 308b648...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3068)

huonw commented on src/librustc_driver/target_features.rs:
```diff
@@ -9,79 +9,53 @@
-pub fn has_avx2(sess: &Session) -> bool {
-    features_contain(sess, "+avx2")
-}
+    let x86_whitelist = [
```
These could be `const X86_WHITELIST: &'static [&'static str] = &["..."];` instead of locals (may neaten the code slightly, since it should avoid the need for `&...[..]` below too).

alexcrichton commented:
@bors: retry

On Tuesday, April 19, 2016, bors <notifications@github.com> wrote:

> [image: :broken_heart:] Test failed - auto-win-msvc-32-opt
> <http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3068>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/31709#issuecomment-212190433>
>

huonw commented on src/librustc_driver/target_features.rs:
```diff
@@ -9,79 +9,53 @@
+
+    let tf = InternedString::new("target_feature");
+    for feat in whitelist {
+        if unsafe { LLVMRustHasFeature(target_machine, feat.as_ptr() as *const c_char) } {
```
It might be nice to have an assertion that the strings end in `\0` here, to ensure people don't make a mistake when adding features in future.

huonw commented:
Thanks for doing this! (Sorry for the delayed comments, been away.)

bors commented:
:hourglass: Testing commit 1ad8561 with merge 0a0296e...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3071)

ranma42 commented:
Applied the improvements suggested by @huonw .

alexcrichton commented:
@bors: r+ ce99a5e5d8ead9140bb9b48c01f1a200950f8ea3

bors commented:
:hourglass: Testing commit ce99a5e with merge 92e3fb3...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/8838), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/8873), [auto-linux-64-cargotest](http://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/162), [auto-linux-64-debug-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2105), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/8812), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/8800), [auto-linux-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/377), [auto-linux-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/785), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/8795), [auto-linux-64-x-armhf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armhf/builds/381), [auto-linux-64-x-armsf](http://buildbot.rust-lang.org/builders/auto-linux-64-x-armsf/builds/375), [auto-linux-64-x-freebsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-freebsd/builds/386), [auto-linux-64-x-netbsd](http://buildbot.rust-lang.org/builders/auto-linux-64-x-netbsd/builds/374), [auto-linux-cross-opt](http://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2015), [auto-linux-musl-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/3955), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/8845), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/8855), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/8817), [auto-mac-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/778), [auto-mac-ios-opt](http://buildbot.rust-lang.org/builders/auto-mac-ios-opt/builds/972), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/3869), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/3886), [auto-win-gnu-32-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/786), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/3881), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/3859), [auto-win-msvc-32-cross-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-cross-opt/builds/253), [auto-win-msvc-32-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3077), [auto-win-msvc-64-cargotest](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/163), [auto-win-msvc-64-opt](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/3870), [auto-win-msvc-64-opt-mir](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/373), [auto-win-msvc-64-opt-rustbuild](http://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/794)
